src/python/pants/util/objects.py

```diff
@@ -330,37 +310,79 @@ def validate_satisfied_by(self, obj):
       "value {!r} (with type {!r}) must satisfy this type constraint: {!r}."
       .format(obj, type(obj).__name__, self))
 
-  def __hash__(self):
-    return hash((type(self), self._types))
-
-  def __eq__(self, other):
-    return type(self) == type(other) and self._types == other._types
-
   def __ne__(self, other):
     return not (self == other)
 
   def __str__(self):
-    if self._desc:
-      constrained_type = '({})'.format(self._desc)
+    return '{}{}'.format(self._variance_symbol, self._description)
+
+
+class BasicTypeConstraint(TypeConstraint):
```

Comment:
`Basic` isn't really a meaningful name here, and I'm still not entirely sure what it means after reading through the code... Can you either come up with a more descriptive name, or write some pydoc on the class explaining why it exists?

Reply:
Yes, this was the one TODO I didn't mark as such. Thank you for pointing it out!

Reply:
I have changed the name to `TypeOnlyConstraint` and added what I thought was a useful pydoc describing the distinction between it and `TypeConstraint` -- I believe this is a lot more understandable now.